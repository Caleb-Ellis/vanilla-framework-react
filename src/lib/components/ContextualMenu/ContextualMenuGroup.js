import React from 'react';
import PropTypes from 'prop-types';
import getClassName from '../../utils/getClassName';

const ContextualMenuGroup = (props) => {
  const { children, className, ...otherProps } = props;

  const classNames = getClassName({
    [className]: className,
    'p-contextual-menu__group': true,
  }) || undefined;

  const items = React.Children.map(children, (child) => {
    const childClasses = getClassName({
      [child.props.className]: child.props.className,
      'p-contextual-menu__link': true,
    });
    return React.cloneElement(child, {
      className: childClasses,
    });
  });

  return (
    <span className={classNames} {...otherProps}>
      {items}
    </span>
  );
};

ContextualMenuGroup.defaultProps = {
  children: null,
  className: null,
};

ContextualMenuGroup.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

ContextualMenuGroup.displayName = 'ContextualMenuGroup';

export default ContextualMenuGroup;
