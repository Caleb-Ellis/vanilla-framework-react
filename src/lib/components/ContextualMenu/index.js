export { default as ContextualMenu } from './ContextualMenu';
export { default as ContextualMenuDropdown } from './ContextualMenuDropdown';
export { default as ContextualMenuGroup } from './ContextualMenuGroup';
