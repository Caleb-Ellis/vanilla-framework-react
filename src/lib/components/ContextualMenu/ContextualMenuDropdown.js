import React from 'react';
import PropTypes from 'prop-types';
import getClassName from '../../utils/getClassName';

import ContextualMenuGroup from './ContextualMenuGroup';

const ContextualMenuDropdown = (props) => {
  const { children, className, ...otherProps } = props;

  const classNames = getClassName({
    [className]: className,
    'p-contextual-menu__dropdown': true,
  }) || undefined;

  const dropdownItems = React.Children.map(children, (child) => {
    if (child.type !== ContextualMenuGroup) {
      const childClasses = getClassName({
        [child.props.className]: child.props.className,
        'p-contextual-menu__link': true,
      });
      return React.cloneElement(child, {
        className: childClasses,
      });
    }
    return child;
  });

  return (
    <span className={classNames} {...otherProps}>
      { dropdownItems }
    </span>
  );
};

ContextualMenuDropdown.defaultProps = {
  children: null,
  className: null,
};

ContextualMenuDropdown.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

ContextualMenuDropdown.displayName = 'ContextualMenuDropdown';

export default ContextualMenuDropdown;
