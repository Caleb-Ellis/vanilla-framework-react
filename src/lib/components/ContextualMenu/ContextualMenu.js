import React from 'react';
import PropTypes from 'prop-types';
import getClassName from '../../utils/getClassName';

import ContextualMenuDropdown from './ContextualMenuDropdown';

class ContextualMenu extends React.Component {
  constructor() {
    super();

    this.toggleOpen = this.toggleOpen.bind(this);
    this.state = { open: false };
  }

  toggleOpen() {
    this.setState({ open: !this.state.open });
  }

  render() {
    const {
      className, children, id, position, ...otherProps
    } = this.props;

    const classNames = getClassName({
      [className]: className,
      'p-contextual-menu': true,
      [`p-contextual-menu--${position}`]: position,
    }) || undefined;

    const toggle = React.Children.map(children, (child) => {
      if (child.type !== ContextualMenuDropdown) {
        return React.cloneElement(child, {
          className: 'p-contextual-menu__toggle',
          'aria-controls': id,
          'aria-expanded': this.state.open,
          'aria-haspopup': 'true',
          role: 'link',
          onClick: this.toggleOpen,
          tabIndex: 0,
        });
      } else if (child.type === ContextualMenuDropdown) {
        return React.cloneElement(child, {
          id,
          'aria-hidden': !this.state.open,
        });
      }
      return child;
    });

    return (
      <span
        className={classNames}
        {...otherProps}
      >
        {toggle}
      </span>
    );
  }
}

ContextualMenu.defaultProps = {
  children: null,
  className: null,
  position: null,
};

ContextualMenu.propTypes = {
  children: (props, propName, componentName) => {
    const prop = props[propName];
    let error = null;
    let count = 0;

    React.Children.forEach(prop, (child) => {
      if (child.type === ContextualMenuDropdown) {
        count += 1;
      }
    });

    if (count !== 1) {
      error = new Error(`${componentName} should have exactly one "ContextualMenuDropdown" child.`);
    }

    return error;
  },
  className: PropTypes.string,
  id: PropTypes.string.isRequired,
  position: PropTypes.oneOf(['left', 'center']),
};

ContextualMenu.displayName = 'ContextualMenu';

export default ContextualMenu;
