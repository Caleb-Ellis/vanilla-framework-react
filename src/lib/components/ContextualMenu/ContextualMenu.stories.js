import React from 'react';
import { storiesOf } from '@storybook/react';
import { select } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';

import ContextualMenu from './ContextualMenu';
import ContextualMenuDropdown from './ContextualMenuDropdown';
import ContextualMenuGroup from './ContextualMenuGroup';
import Button from '../Button/Button';
import Link from '../Link/Link';

storiesOf('ContextualMenu', module)
  .add('Default',
    withInfo('A ContextualMenu component can be wrapped around any component (button, link, navigation item etc.) to give it a secondary dropdown menu on click. One of the children must be a ContextualMenuDropdown that contains the links (or other elements) you would like to display. ContextualMenus must be given an "id" prop so the dropdown knows where in the DOM to appear, and it also accepts a "position" prop to determine which side of the element appears (default "right").')(() => (
      <div style={{ paddingLeft: '5rem' }}>
        <ContextualMenu id="menu" position={select('Position', [undefined, 'left', 'center'], undefined)}>
          <Button neutral>Menu</Button>
          <ContextualMenuDropdown>
            <Link href="#a">Link 1</Link>
            <Link href="#a">Link 2</Link>
            <Link href="#a">Link 3</Link>
          </ContextualMenuDropdown>
        </ContextualMenu>
      </div>),
    ),
  )

  .add('Groups',
    withInfo('Dropdown links can be grouped when placed inside a ContextualMenuGroup component, which separates groups by a dividing line.')(() => (
      <div style={{ paddingLeft: '5rem' }}>
        <ContextualMenu id="menu" position={select('Position', [undefined, 'left', 'center'], undefined)}>
          <Button neutral>Groups</Button>
          <ContextualMenuDropdown>
            <ContextualMenuGroup>
              <Link href="#a">Link 1</Link>
              <Link href="#a">Link 2</Link>
            </ContextualMenuGroup>
            <ContextualMenuGroup>
              <Link href="#a">Link 3</Link>
              <Link href="#a">Link 4</Link>
            </ContextualMenuGroup>
          </ContextualMenuDropdown>
        </ContextualMenu>
      </div>),
    ),
  )

  .add('Link',
    withInfo('ContextualMenus can be applied to any element, including links within paragraph text.')(() => (
      <p>
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Modi natus atque eligendi
        deleniti hic, dolores veritatis reiciendis officiis illo, facere facilis accusamus
        similique nulla nesciunt.&nbsp;
        <ContextualMenu id="menu" position={select('Position', [undefined, 'left', 'center'], undefined)}>
          <Link href="#a" onClick={() => null}>Nemo</Link>
          <ContextualMenuDropdown>
            <ContextualMenuGroup>
              <Link href="#a">Learn more</Link>
              <Link href="#a">Learn more</Link>
            </ContextualMenuGroup>
            <ContextualMenuGroup>
              <Link href="#a">Home</Link>
            </ContextualMenuGroup>
          </ContextualMenuDropdown>
        </ContextualMenu>
        &nbsp;reprehenderit officia assumenda error enim. Recusandae
        reiciendis ipsum, mollitia illo iusto excepturi alias dolore fugit eligendi nostrum, unde
        architecto consequuntur similique quo! Maxime iusto facere, commodi iste fuga officiis.
      </p>),
    ),
  );
