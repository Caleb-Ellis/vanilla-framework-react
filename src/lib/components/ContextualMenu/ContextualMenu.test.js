import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import ContextualMenu from './ContextualMenu';
import ContextualMenuDropdown from './ContextualMenuDropdown';
import ContextualMenuGroup from './ContextualMenuGroup';

describe('<ContextualMenu>', () => {
  it('should render a simple ContextualMenu correctly', () => {
    const menu = ReactTestRenderer.create(
      <ContextualMenu id="menu">
        <div>Menu</div>
        <ContextualMenuDropdown>
          <a href="#a">Link 1</a>
          <a href="#a">Link 2</a>
          <a href="#a">Link 3</a>
        </ContextualMenuDropdown>
      </ContextualMenu>);
    const json = menu.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render with position prop correctly', () => {
    const menu = ReactTestRenderer.create(
      <div>
        <ContextualMenu id="menu" position="left">
          <div>Menu</div>
          <ContextualMenuDropdown>
            <a href="#a">Link 1</a>
            <a href="#a">Link 2</a>
            <a href="#a">Link 3</a>
          </ContextualMenuDropdown>
        </ContextualMenu>
        <ContextualMenu id="menu" position="center">
          <div>Menu</div>
          <ContextualMenuDropdown>
            <a href="#a">Link 1</a>
            <a href="#a">Link 2</a>
            <a href="#a">Link 3</a>
          </ContextualMenuDropdown>
        </ContextualMenu>
      </div>);
    const json = menu.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render ContextualMenuGroups correctly', () => {
    const menu = ReactTestRenderer.create(
      <ContextualMenu id="menu" position="left">
        <div>Menu</div>
        <ContextualMenuDropdown>
          <ContextualMenuGroup>
            <a href="#a">Link 1</a>
            <a href="#a">Link 2</a>
          </ContextualMenuGroup>
          <ContextualMenuGroup>
            <a href="#a">Link 3</a>
            <a href="#a">Link 4</a>
          </ContextualMenuGroup>
        </ContextualMenuDropdown>
      </ContextualMenu>);
    const json = menu.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should open the ContextualMenu when clicked', () => {
    const menu = shallow(
      <ContextualMenu id="menu">
        <button>Menu</button>
        <ContextualMenuDropdown>
          <a href="#a">Link 1</a>
          <a href="#a">Link 2</a>
          <a href="#a">Link 3</a>
        </ContextualMenuDropdown>
      </ContextualMenu>);

    expect(menu.find('[aria-expanded=true]')).toHaveLength(0);
    expect(menu.find('[aria-hidden=true]')).toHaveLength(1);

    menu.find('button').hostNodes().simulate('click');
    expect(menu.find('[aria-expanded=true]')).toHaveLength(1);
    expect(menu.find('[aria-hidden=true]')).toHaveLength(0);
  });
});
