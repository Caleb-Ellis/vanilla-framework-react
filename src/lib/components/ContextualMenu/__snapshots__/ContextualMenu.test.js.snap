// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContextualMenu> should render ContextualMenuGroups correctly 1`] = `
<span
  className="p-contextual-menu p-contextual-menu--left"
>
  <div
    aria-controls="menu"
    aria-expanded={false}
    aria-haspopup="true"
    className="p-contextual-menu__toggle"
    onClick={[Function]}
    role="link"
    tabIndex={0}
  >
    Menu
  </div>
  <span
    aria-hidden={true}
    className="p-contextual-menu__dropdown"
    id="menu"
  >
    <span
      className="p-contextual-menu__group"
    >
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 1
      </a>
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 2
      </a>
    </span>
    <span
      className="p-contextual-menu__group"
    >
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 3
      </a>
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 4
      </a>
    </span>
  </span>
</span>
`;

exports[`<ContextualMenu> should render a simple ContextualMenu correctly 1`] = `
<span
  className="p-contextual-menu"
>
  <div
    aria-controls="menu"
    aria-expanded={false}
    aria-haspopup="true"
    className="p-contextual-menu__toggle"
    onClick={[Function]}
    role="link"
    tabIndex={0}
  >
    Menu
  </div>
  <span
    aria-hidden={true}
    className="p-contextual-menu__dropdown"
    id="menu"
  >
    <a
      className="p-contextual-menu__link"
      href="#a"
    >
      Link 1
    </a>
    <a
      className="p-contextual-menu__link"
      href="#a"
    >
      Link 2
    </a>
    <a
      className="p-contextual-menu__link"
      href="#a"
    >
      Link 3
    </a>
  </span>
</span>
`;

exports[`<ContextualMenu> should render with position prop correctly 1`] = `
<div>
  <span
    className="p-contextual-menu p-contextual-menu--left"
  >
    <div
      aria-controls="menu"
      aria-expanded={false}
      aria-haspopup="true"
      className="p-contextual-menu__toggle"
      onClick={[Function]}
      role="link"
      tabIndex={0}
    >
      Menu
    </div>
    <span
      aria-hidden={true}
      className="p-contextual-menu__dropdown"
      id="menu"
    >
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 1
      </a>
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 2
      </a>
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 3
      </a>
    </span>
  </span>
  <span
    className="p-contextual-menu p-contextual-menu--center"
  >
    <div
      aria-controls="menu"
      aria-expanded={false}
      aria-haspopup="true"
      className="p-contextual-menu__toggle"
      onClick={[Function]}
      role="link"
      tabIndex={0}
    >
      Menu
    </div>
    <span
      aria-hidden={true}
      className="p-contextual-menu__dropdown"
      id="menu"
    >
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 1
      </a>
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 2
      </a>
      <a
        className="p-contextual-menu__link"
        href="#a"
      >
        Link 3
      </a>
    </span>
  </span>
</div>
`;
